### ISSUE TITLE
HEAD of OpenSSL_1_1_0-stable crashes on macOS when building with "no-dso"

### ISSUE THREAD

**Issue by smuehlst** [openssl/openssl#6493]
I'm porting an application from OpenSSL 1.0.2 to OpenSSL 1.1.0. Because I need commit 6912debb881e669f7a7fb621588e20347111c4f0 "Add APIs for custom X509_LOOKUP_METHOD creation" I'm working with the current HEAD of the OpenSSL_1_1_0-stable branch, concretely with 0c27d793745c7837b13646302b6890a556b7017a.

I'm experiencing an immediate crash with any application that calls OPENSSL_init_crypto() when I build on macOS with the `config` options `no-shared` and `no-dso`.

How to reproduce:
```
$ ./config no-shared no-dso
Operating system: x86_64-apple-darwinDarwin Kernel Version 16.7.0: Fri Apr 27 17:59:46 PDT 2018; root:xnu-3789.73.13~1/RELEASE_X86_64
WARNING! If you wish to build 32-bit library, then you have to
         invoke 'KERNEL_BITS=32 ./config  no-shared no-dso'.
         You have about 5 seconds to press Ctrl-C to abort.
Configuring for darwin64-x86_64-cc
Configuring OpenSSL version 1.1.0i-dev (0x10100090L)
    no-asan         [default]  OPENSSL_NO_ASAN
    no-crypto-mdebug [default]  OPENSSL_NO_CRYPTO_MDEBUG
    no-crypto-mdebug-backtrace [default]  OPENSSL_NO_CRYPTO_MDEBUG_BACKTRACE
    no-dso          [option]  
    no-dynamic-engine [forced]  
    no-ec_nistp_64_gcc_128 [default]  OPENSSL_NO_EC_NISTP_64_GCC_128
    no-egd          [default]  OPENSSL_NO_EGD
    no-fuzz-afl     [default]  OPENSSL_NO_FUZZ_AFL
    no-fuzz-libfuzzer [default]  OPENSSL_NO_FUZZ_LIBFUZZER
    no-heartbeats   [default]  OPENSSL_NO_HEARTBEATS
    no-md2          [default]  OPENSSL_NO_MD2 (skip dir)
    no-msan         [default]  OPENSSL_NO_MSAN
    no-rc5          [default]  OPENSSL_NO_RC5 (skip dir)
    no-sctp         [default]  OPENSSL_NO_SCTP
    no-shared       [option]  
    no-ssl-trace    [default]  OPENSSL_NO_SSL_TRACE
    no-ssl3         [default]  OPENSSL_NO_SSL3
    no-ssl3-method  [default]  OPENSSL_NO_SSL3_METHOD
    no-ubsan        [default]  OPENSSL_NO_UBSAN
    no-unit-test    [default]  OPENSSL_NO_UNIT_TEST
    no-weak-ssl-ciphers [default]  OPENSSL_NO_WEAK_SSL_CIPHERS
    no-zlib         [default] 
    no-zlib-dynamic [default] 
Configuring for darwin64-x86_64-cc
CC            =cc
CFLAG         =-O3 -D_REENTRANT -arch x86_64 -DL_ENDIAN -Wall 
SHARED_CFLAG  =-fPIC
DEFINES       =NDEBUG OPENSSL_THREADS OPENSSL_NO_DYNAMIC_ENGINE OPENSSL_PIC OPENSSL_IA32_SSE2 OPENSSL_BN_ASM_MONT OPENSSL_BN_ASM_MONT5 OPENSSL_BN_ASM_GF2m SHA1_ASM SHA256_ASM SHA512_ASM RC4_ASM MD5_ASM AES_ASM VPAES_ASM BSAES_ASM GHASH_ASM ECP_NISTZ256_ASM PADLOCK_ASM POLY1305_ASM
LFLAG         =
PLIB_LFLAG    =-Wl,-search_paths_first
EX_LIBS       = 
APPS_OBJ      =
CPUID_OBJ     =x86_64cpuid.o
UPLINK_OBJ    =
BN_ASM        =asm/x86_64-gcc.o x86_64-mont.o x86_64-mont5.o x86_64-gf2m.o rsaz_exp.o rsaz-x86_64.o rsaz-avx2.o
EC_ASM        =ecp_nistz256.o ecp_nistz256-x86_64.o
DES_ENC       =des_enc.o fcrypt_b.o
AES_ENC       =aes-x86_64.o vpaes-x86_64.o bsaes-x86_64.o aesni-x86_64.o aesni-sha1-x86_64.o aesni-sha256-x86_64.o aesni-mb-x86_64.o
BF_ENC        =bf_enc.o
CAST_ENC      =c_enc.o
RC4_ENC       =rc4-x86_64.o rc4-md5-x86_64.o
RC5_ENC       =rc5_enc.o
MD5_OBJ_ASM   =md5-x86_64.o
SHA1_OBJ_ASM  =sha1-x86_64.o sha256-x86_64.o sha512-x86_64.o sha1-mb-x86_64.o sha256-mb-x86_64.o
RMD160_OBJ_ASM=
CMLL_ENC      =cmll-x86_64.o cmll_misc.o
MODES_OBJ     =ghash-x86_64.o aesni-gcm-x86_64.o
PADLOCK_OBJ   =e_padlock-x86_64.o
CHACHA_ENC    =chacha-x86_64.o
POLY1305_OBJ  =poly1305-x86_64.o
BLAKE2_OBJ    =
PROCESSOR     =
RANLIB        =ranlib -c
ARFLAGS       =
PERL          =/opt/local/bin/perl5.24

SIXTY_FOUR_BIT_LONG mode

Configured for darwin64-x86_64-cc.
$ make build_programs
<make output omitted>
$ apps/openssl version
Illegal instruction: 4
```
The crash does not occur when doing the same build with OpenSSL_1_1_0h.
`git bisect` identifies the following bad commit:
```
bf21fe935a979c08292d06553ef8c9a49382208c is the first bad commit
commit bf21fe935a979c08292d06553ef8c9a49382208c
Author: Bernd Edlinger <bernd.edlinger@hotmail.de>
Date:   Sat Apr 21 15:41:42 2018 +0200

    Ensure the thread keys are always allocated in the same order
    
    Back-port of #5911
    Fixes: #5899
    
    Reviewed-by: Rich Salz <rsalz@openssl.org>
    (Merged from https://github.com/openssl/openssl/pull/6037)
```
This happens on macOS Sierra with the following compiler version:
```
$ cc --version
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by mattcaswell** [comment]
Ping @bernd-edlinger. Thoughts?

**Comment by mattcaswell** [comment]
Does it only happen with that combination of options? i.e. if you drop one or the other or both does it still fail?

**Comment by smuehlst** [comment]
> Does it only happen with that combination of options? i.e. if you drop one or the other or both does it still fail?

The problematic option is `no-dso`. If I leave that out the `openssl` binary does not crash. `no-shared` is necessary because I want a statically linked `openssl` binary, because this is the same situation with my application that links OpenSSL statically.

But the test case can indeed be reduced by leaving out the `no-shared` option and by calling the dynamically linked `openssl`binary like this:
```
$ ./config no-dso
<config output omitted>
$ make build_programs
<make output omitted>
$ DYLD_LIBRARY_PATH=$(pwd) apps/openssl version
Illegal instruction: 4
```

**Comment by bernd-edlinger** [comment]
you could try to build with -DOPENSSL_INIT_DEBUG to get some traces from the initialization
code, and maybe with -d to get debug info.
A call stack where this crash happens would be helpful.

**Comment by mattcaswell** [comment]
I can manage to get linux to show a problem with no-dso on the head of the 1.1.0 branch (but not master), if I apply this patch:

````
diff --git a/Configurations/10-main.conf b/Configurations/10-main.conf
index 7e66705261..1154d84a97 100644
--- a/Configurations/10-main.conf
+++ b/Configurations/10-main.conf
@@ -627,8 +627,7 @@ sub vms_info {
         thread_scheme    => "pthreads",
         dso_scheme       => "dlfcn",
         shared_target    => "linux-shared",
-        shared_cflag     => "-fPIC -DOPENSSL_USE_NODELETE",
-        shared_ldflag    => "-Wl,-znodelete",
+        shared_cflag     => "-fPIC",
         shared_extension => ".so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)",
     },
     "linux-generic64" => {
````

Building the 1.1.0 branch with the above applied and using the no-dso option results in a hang (not a crash) from commit bf21fe935a97, but not before it. I assume this is the same problem but manifesting itself differently on Linux vs MacOS.

**Comment by bernd-edlinger** [comment]
Thanks Matt!
While GNU/Linux hangs on recursive run once functions, MacOS returns a failure.
From the call stack it seems that ossl_init_load_crypto_nodelete calls DSO_dsobyaddr
this calls again ERR_put_error.  That master works fine means probably that I need
to back-port at least the patch which introduced the err_shelve_state.

**Comment by mattcaswell** [comment]
> While GNU/Linux hangs on recursive run once functions, MacOS returns a failure.
From the call stack it seems that ossl_init_load_crypto_nodelete calls DSO_dsobyaddr
this calls again ERR_put_error. That master works fine means probably that I need
to back-port at least the patch which introduced the err_shelve_state.

Great!

Interestingly I noticed that the `OPENSSL_NO_DSO` guards in init.c do nothing at all. That macro never gets defined. Instead if you ask for "no-dso" you still get all the DSO symbols, its just that you get the "null" DSO method which doesn't do anything. So what is happening here is that we are entering the `ossl_init_load_crypto_nodelete` run once code, going down the DSO path, and then encountering an error which ultimately results in an `ERR_put_error call`.

**Comment by bernd-edlinger** [comment]
It's 6b49b30811f4afa0340342af9400b8d0357b5291 does the trick.
But probably the NO_DSO needs to be fixed as well.

**Comment by bernd-edlinger** [comment]
#6494

**Comment by smuehlst** [comment]
I've rebuilt the `openssl` executable after configuring with `./config --debug -DOPENSSL_INIT_DEBUG no-shared no-dso`, then invoked the `openssl` executable under lldb: 
```
$ lldb apps/openssl
(lldb) target create "apps/openssl"
Current executable set to 'apps/openssl' (x86_64).
(lldb) r version
Process 22123 launched: '/Users/stm/git/openssl/apps/openssl' (x86_64)
OPENSSL_INIT: ossl_init_base: Setting up stop handlers
OPENSSL_INIT: ossl_init_load_crypto_nodelete()
Process 22123 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x00007fff96cffc6e libsystem_platform.dylib`_os_once_gate_recursive_abort + 23
libsystem_platform.dylib`_os_once_gate_recursive_abort:
->  0x7fff96cffc6e <+23>: ud2    

libsystem_platform.dylib`_os_once_gate_corruption_abort:
    0x7fff96cffc70 <+0>:  movl   %edi, %eax
    0x7fff96cffc72 <+2>:  leaq   0x207f(%rip), %rcx        ; "BUG IN CLIENT OF LIBPLATFORM: os_once_t is corrupt"
    0x7fff96cffc79 <+9>:  movq   %rcx, 0x8d03640(%rip)     ; gCRAnnotations + 8
Target 0: (openssl) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
  * frame #0: 0x00007fff96cffc6e libsystem_platform.dylib`_os_once_gate_recursive_abort + 23
    frame #1: 0x00007fff96cff984 libsystem_platform.dylib`_os_once_gate_wait_slow + 111
    frame #2: 0x00007fff96d06b2b libsystem_pthread.dylib`pthread_once + 57
    frame #3: 0x000000010026470d openssl`CRYPTO_THREAD_run_once(once=0x00000001003340c0, init=(openssl`ossl_init_load_crypto_nodelete_ossl_ at init.c:103)) at threads_pthread.c:106
    frame #4: 0x00000001001ef4c1 openssl`OPENSSL_init_crypto(opts=0, settings=0x0000000000000000) at init.c:548
    frame #5: 0x00000001001c272a openssl`ERR_get_state at err.c:670
    frame #6: 0x00000001001c250e openssl`ERR_put_error(lib=37, func=105, reason=108, file="crypto/dso/dso_lib.c", line=315) at err.c:364
    frame #7: 0x000000010016f382 openssl`DSO_pathbyaddr(addr=0x000000010033c020, path=0x0000000000000000, sz=0) at dso_lib.c:315
    frame #8: 0x000000010016f3e1 openssl`DSO_dsobyaddr(addr=0x000000010033c020, flags=4) at dso_lib.c:325
    frame #9: 0x00000001001f0165 openssl`ossl_init_load_crypto_nodelete at init.c:137
    frame #10: 0x00000001001efd49 openssl`ossl_init_load_crypto_nodelete_ossl_ at init.c:103
    frame #11: 0x00007fff96d06b8c libsystem_pthread.dylib`__pthread_once_handler + 65
    frame #12: 0x00007fff96cfbac1 libsystem_platform.dylib`_os_once + 36
    frame #13: 0x00007fff96d06b2b libsystem_pthread.dylib`pthread_once + 57
    frame #14: 0x000000010026470d openssl`CRYPTO_THREAD_run_once(once=0x00000001003340c0, init=(openssl`ossl_init_load_crypto_nodelete_ossl_ at init.c:103)) at threads_pthread.c:106
    frame #15: 0x00000001001ef4c1 openssl`OPENSSL_init_crypto(opts=0, settings=0x0000000000000000) at init.c:548
    frame #16: 0x00000001001ea047 openssl`do_ex_data_init at ex_data.c:41
    frame #17: 0x00000001001ea019 openssl`do_ex_data_init_ossl_ at ex_data.c:39
    frame #18: 0x00007fff96d06b8c libsystem_pthread.dylib`__pthread_once_handler + 65
    frame #19: 0x00007fff96cfbac1 libsystem_platform.dylib`_os_once + 36
    frame #20: 0x00007fff96d06b2b libsystem_pthread.dylib`pthread_once + 57
    frame #21: 0x000000010026470d openssl`CRYPTO_THREAD_run_once(once=0x00000001003340a0, init=(openssl`do_ex_data_init_ossl_ at ex_data.c:39)) at threads_pthread.c:106
    frame #22: 0x00000001001e9110 openssl`get_and_lock(class_index=12) at ex_data.c:60
    frame #23: 0x00000001001e9517 openssl`CRYPTO_new_ex_data(class_index=12, obj=0x0000000100700200, ad=0x0000000100700260) at ex_data.c:215
    frame #24: 0x00000001000ffa81 openssl`BIO_new(method=0x0000000100301ce8) at bio_lib.c:29
    frame #25: 0x00000001001054ac openssl`BIO_new_fp(stream=0x00007fff9f9e8110, close_flag=16) at bss_file.c:101
    frame #26: 0x0000000100002d9e openssl`dup_bio_in(format=32769) at apps.c:2388
    frame #27: 0x0000000100024e6b openssl`main(argc=2, argv=0x00007fff5fbff838) at openssl.c:127
    frame #28: 0x0000000100527235 libdyld.dylib`start + 1
    frame #29: 0x0000000100527235 libdyld.dylib`start + 1
(lldb
```

**Comment by bernd-edlinger** [comment]
PR #6494 should fix this.

**Comment by mattcaswell** [comment]
I raised issue #6495 for the `OPENSSL_NO_DSO` isn't defined problem.

**Comment by bernd-edlinger** [comment]
The crash is fixed by 6b49b30, the remaining issue is tracked in #6495.
Thanks for reporting this issue!
